src/jni_core/src/lib.rs

```diff
@@ -152,21 +152,36 @@ impl<T> From<T> for Pointer<'static, T> {
 
 impl<'a, T> Pointer<'a, T> {
     fn as_ref(&self) -> &'a T {
-        debug_assert!(self.pointer != 0);
+        assert!(self.pointer != 0);
         unsafe { &*(self.pointer as *const T) }
     }
 
     fn as_mut(&mut self) -> &'a mut T {
-        debug_assert!(self.pointer != 0);
+        assert!(self.pointer != 0);
         unsafe { &mut *(self.pointer as *mut T) }
     }
 }
 
+/// A pointer that owns the object it points to.
+///
+/// Note that dropping an `OwnedPointer` does not release the object.
+/// Instead, you should call [`OwnedPointer::release`] manually.
 pub type OwnedPointer<T> = Pointer<'static, T>;
 
 impl<T> OwnedPointer<T> {
-    fn drop(self) {
-        debug_assert!(self.pointer != 0);
+    /// Consume `self` and return the pointer value. Used for passing to JNI.
+    pub fn into_pointer(self) -> jlong {
+        self.pointer
+    }
+
+    /// Release the object behind the pointer.
+    fn release(self) {
+        tracing::debug!(
+            type_name = std::any::type_name::<T>(),
+            address = self.pointer,
+            "release jni OwnedPointer"
+        );
```

Comment:
I'm not able to observe this line in the log. Not sure if I'm using `ref.Cleaner` correctly...

Reply:
Not sure if it's because memory is enough and GC is not triggered. You can try running `jcmd <pid> GC.run` to trigger the gc.

Reply:
I realized that the channel is considered reachable by Java even if the Debezium runner is stopped, since we forget to call `JniDbzSourceRegistry.unregister` in the normal path. Add a commit to move it into the `finally` block and triggering GC with `jcmd` now successfully releases the memory.

```
2025-02-11T17:53:06.507871+08:00 DEBUG risingwave_jni_core: release jni OwnedPointer type_name="tokio::sync::mpsc::bounded::Sender<core::result::Result<risingwave_pb::connector_service::GetEventStreamResponse, anyhow::Error>>" address=5795941504
```

<img width="612" alt="image" src="https://github.com/user-attachments/assets/4e43e3d3-6228-444e-8ae2-81f6fa13c7ae" />